Explicit Codable & CodingKey Derived Conformance

itaiferber opened the pull request:
**What's in this pull request?**
This PR integrates preliminary compiler work to support derived conformance for the newly introduced `Codable` and `CodingKey` types which are part of #8124.

This PR focuses on internal compiler features for improving the experience of using the features provided in #8124.

A swift-evolution proposal will be introduced momentarily and will link to this PR.

**NOTE:** This PR should _not_ be merged until #8124 is.

This work is **not** complete. Remaining top-level features which will be iterated upon:
- [x] Need to not generate `decode` expressions for immutable vars with default initializers
- [ ] Provide failure diagnostics
- [ ] Add unit tests:
    - [ ] Imported types
    - [ ] Collections (when possible)
    - [ ] Generic types
    - [ ] Nested types
- [ ] Synthesize `CodingKeys` type even w/o other derivation

Commit:
Explicitly derive Codable & CodingKey conformance

Integrate synthesis of Codable and CodingKey requirements along with preliminary unit tests

slavapestov commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,949 @@
+
+  auto funcDC = cast<DeclContext>(encodeDecl);
+  ASTContext &C = funcDC->getASTContext();
+  auto &tc = *(TypeChecker *)C.getLazyResolver();
```
Please just pass down a TypeChecker instead of doing an unsafe cast here

itaiferber replied:
Unfortunately, we can't. 😕 This is in the body synthesizer for the `encode(to:)` method, which can only accept the `AbstractFunctionDecl` whose body it's synthesizing.

The only way we can do this more safely that I can tell is to eagerly generate the body directly in `deriveCodable_encode` below, instead of using a body synthesizer. Is that better?

slavapestov replied:
I see. I think the conformance itself is lazily synthesized, so there should not be much of a perf win from lazily synthesizing the bodies. However @jrose-apple should comment on this.

itaiferber replied:
FWIW, I followed the exact model that `RawRepresentable`, `Equatable`, and `Hashable` use. If we eagerly synthesize the body here we should probably do it for those as well.

slavapestov replied:
Those don't need to cast the resolver to a TypeChecker though...

jrose-apple replied:
I don't think synthesizing the body should emit diagnostics. All of that checking should be done up front like it is for the other synthesized conformance members. Without that, I don't think you would need a TypeChecker.

itaiferber replied:
I guess a refactor might be in order here. The `TypeChecker` is only used to be passed into `getOrSynthesizeCodingKeysDecl`, which at this point should only really be performing a lookup (because it would have already been derived here otherwise). Thinking about it, this is even strictly unnecessary; I should be able to just `lookupDirect` the `CodingKeys` decl since at this point it should be guaranteed to be there.

itaiferber replied:
I realized the type checker is actually used further down to tell if the type's superclass is `Codable`, so we can't get rid of it. Should we not synthesize the body lazily, then?

slavapestov replied:
If you're just doing a conformance check and not trying to record a dependency or emit a diagnostic, you can call ModuleDecl::lookupConformance() instead of TypeChecker::conformsToProtocol().

itaiferber replied:
`ModuleDecl::lookupConformance` requires a lazy resolver: ` Optional<ProtocolConformanceRef>
  lookupConformance(Type type, ProtocolDecl *protocol, LazyResolver *resolver)`
Should be safe to pass `C.getLazyResolver()` without a cast then, yeah?

itaiferber replied:
Doing this allows me to get rid of `tc` altogether:

```swift
static bool superclassIsCodable(ClassDecl *type) {
  assert(type && "Cannot get superclass of null type.");
  if (!type->hasSuperclass())
    return false;

  auto &C = type->getASTContext();
  auto *codableProto = C.getProtocol(KnownProtocolKind::Codable);

  auto *superclassDecl = type->getSuperclassDecl();
  auto *superclassModule = superclassDecl->getModuleContext();
  return (bool)superclassModule->lookupConformance(type->getSuperclass(),
                                                   codableProto,
                                                   C.getLazyResolver());
}
```

slavapestov commented on test/decl/protocol/special/coding/Inputs/class_codable_simple_multi1.swift:
```diff
@@ -0,0 +1,25 @@
+// REQUIRES: objc_interop
```
Do these really require objc interop or does this all work on Linux with swift-corelibs-foundation too?

itaiferber replied:
These require the import of Foundation, which necessitates the inclusion of `objc_interop`, no?

slavapestov replied:
On Linux, objc_interop is off, but you might still have a Foundation module (swift-corelibs-foundation is pure Swift and doesn't use the ObjC runtime as you're no doubt aware).

I think tests should not pull in Foundation on Linux though, however we might still be able to test the conformance derivation by cooking up a 'mock' Foundation module that only provides the necessary protocols and nothing else.

itaiferber replied:
Doug mentioned this too. Keep in mind though that these protocols use `Data`, which may pull in other things which really do require ObjC; this may prohibit us from being able to mock this in a truly `objc_interop`-free way...

slavapestov replied:
So to clarify -- this code does work on Linux with corelibs-foundation right? It's just that it's hard to test in this scenario?

Would adding some tests that rely on the conformance being derived to swift-corelibs-foundation be an acceptable solution then?

itaiferber replied:
Yes, I believe so. swift-corelibs-foundation will require some changes to bring it in line with the implementation of `JSONEncoder` and `JSONDecoder` specifically, but the primary issue in testing is not one of actual ObjC interop, but one of the testing framework.

All of my implementations should be able to land in swift-corelibs-foundation as well with no changes.

slavapestov commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,949 @@
+// Returns whether the type represented by the given ClassDecl inherits from a
+// type which is Codable.
+static bool classInheritsCodable(TypeChecker &tc, ClassDecl *type) {
+  if (!type || !type->hasSuperclass())
```
What if the class conforms to Codable but has no superclass?

itaiferber replied:
That's handled by default below in `getOrSynthesizeCodingKeysDecl`, `deriveCodable_encode`, and `deriveCodable_init`. This is used to determine whether a class has a `Codable` superclass so we can emit a `super` key, and call `try super.encode(to: container.superEncoder())` and `try super.init(from: container.superDecoder())`; if the class is `Codable` but its superclass is not, we don't need to emit those things.

Perhaps a better name for this function would just be `superclassIsCodable` or similar?

slavapestov replied:
Yes, that sounds like a better name. Also, can you remove the '!type' check or replace it with an assertion? It's not clear to me that this can happen and we should not let it happen silently if it does.

itaiferber replied:
Yes, I can remove that; there's no real need for it.

slavapestov commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,949 @@
+    //
+    // Validate the decl eagerly.
+    if (!varDecl->hasType())
+      tc.validateDecl(varDecl);
```
hasType/getType will assert if the VarDecl is in a different module. This can happen if you define the Codable conformance in an extension. Can you test this case?

You should use getInterfaceType() instead

itaiferber replied:
I caught this case in testing exactly that — asking the `TypeChecker` to validate the decl eagerly evaluates the type so that past the `evaluateDecl` line, `hasType() == true`. However, if `getInterfaceType()` allows us to avoid doing this eager evaluation, then we should use that. Looking into it now.

slavapestov replied:
You should not call validateDecl() on deserialized decks. They are already fully validated

itaiferber replied:
`validateDecl()` is the only way I see that we can do this. `getInterfaceType()` also asserts, since the decl has no interface type set either:

    Assertion failed: (hasInterfaceType() && "No interface type was set"), function getInterfaceType, file /Users/itai/Development/swift-oss/swift/lib/AST/Decl.cpp, line 1769.
error: Execution was interrupted, reason: signal SIGABRT.

slavapestov replied:
validateDecl() is the right thing to call if the decl is in the same module and has not yet been type checked.

itaiferber replied:
What's the solution here, then? The var decl has no type, nor an interface type, but we need the type to be evaluated in order to continue.

slavapestov commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,949 @@
+  // Bind Keyed*Container to Keyed*Container<KeyType>
+  Type boundType[1] = { keyType };
+  auto containerType = BoundGenericType::get(keyedContainerDecl, Type(),
+                                             C.AllocateCopy(boundType));
```
I don't think the AllocateCopy is necessary

itaiferber replied:
I think it is — `BoundGenericType::get` takes an `ArrayRef` referencing the bound types, which it does not copy. The `ArrayRef` will then be pointing to whatever it was created with, in this case, an array on the stack. We need to allocate a copy or else we'll have a dangling pointer. No?

slavapestov replied:
You're right. It's hard to remember which ArrayRefs are copied and which ones are not. It's a bit of a mess...

itaiferber replied:
Don't worry, I discovered this myself — lots of fun trying to diagnose the undefined behavior... ;)

gottesmm replied:
In general in these cases, the right thing to do is to use a SmallVector or a TinyPtrVector.

itaiferber replied:
A `SmallVector` is stack-allocated, though, and the `ArrayRef` (which is not copied in the cases here) will keep pointing to the `SmallVector` on the stack long past the `SmallVector` going away.

itaiferber replied:
I originally used `SmallVector`s all over the place and was seeing strange behavior, until @DougGregor corrected me.

gottesmm replied:
So, we are storing the ArrayRef here? Sorry I did a drive by. This brings up an interesting question though. If everyone is already copying here, it may make sense to just change this API to always copy. Internally if we run into overhead, we can optimize later. I am not saying that you should fix this in this PR. Just speaking out loud.

itaiferber replied:
`BoundGenericType::get` takes an `ArrayRef` without copying it. Most decls do, in fact, because they likely assume that these things were already allocated earlier in the parsing process (which they were), so it's safe to take a reference without copying. Here, however, I'm synthesizing it all from scratch, so it does need to be allocated because these references are long-lived.

gottesmm replied:
I understand. What I was saying is that in the past this has been a source of bugs. Having a more conservative API here from a memory management perspective would be good since those bugs would be defined away. Like I said before, I do not think that this PR is the place to make such a change and I am not putting it on you to fix.

itaiferber replied:
Yep! We're not in disagreement here 😄 An audit of those would be a good idea (some are even taken as `MutableArrayRef`s...), but not part of this.

slavapestov reviewed (commented):
Nice work overall, I only have a few nitpicks so far.

Can you add some tests for generic types? I only see non generics here.

itaiferber commented:
@slavapestov Thanks for the review! As for the generic tests, absolutely — that's on the to-do list above.

slavapestov commented:
I think you can merge this whenever you're ready -- we can sort out the remaining issues later. I hope you get some time to work on the various TODO comments too :)

itaiferber commented:
@slavapestov It seems I don't have write access for this repo — do you mind running a CI test please?

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
Wait, don't we have to merge the other PR first?

itaiferber commented:
I don't want to merge this, though, until #8124 lands as well

itaiferber commented:
Yeah, we definitely shouldn't merge. Does the smoke test run the unit tests? Those will fail because the `Codable` and `CodingKey` protocols won't actually be in Foundation...

slavapestov commented:
Yes, the smoke tests run the tests, that's the whole point of presubmit testing :)

itaiferber commented:
Duh, yeah. Then this will almost certainly fail ¯\_(ツ)_/¯

Commit:
classInheritsCodable → superclassIsCodable

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+#include "TypeChecker.h"
```
This needs a swift header.

itaiferber replied:
What do you mean, like `#include "swift/.../TypeChecker.h"`? This is not a public header; it is in `lib/Sema/TypeChecker.h`, right alongside `lib/Sema/DerivedConformanceCodable.cpp`. This is how all other clients using `TypeChecker` import it.

itaiferber replied:
Adding this

gottesmm commented on lib/Sema/DerivedConformanceCodingKey.cpp:
```diff
@@ -0,0 +1,615 @@
+#include "TypeChecker.h"
```
This needs a swift header.

itaiferber replied:
Same here.

itaiferber replied:
Adding this

gottesmm reviewed (commented):
@itaiferber I am looking through this for swift style violations...

I am seeing a couple already. Have you run this through clang-format? It will lower the amount of style violations I need to comment on.

gottesmm commented on lib/AST/Decl.cpp:
```diff
@@ -2092,9 +2092,39 @@ bool NominalTypeDecl::derivesProtocolConformance(ProtocolDecl *protocol) const {
     default:
       return false;
     }
+  } else if (isa<StructDecl>(this) || isa<ClassDecl>(this)) {
```
This violates LLVM style:

http://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

All of the code paths before the else if return. The else is not necessary. In fact, you could even invert the if condition and do an early return false if you want to.

gottesmm replied:
When you fix this, can you also add a * to the auto above on knownProtocol. (Or even change it to a reference). I know that this is something that is not in the code that you have touched yourself.

itaiferber replied:
Fixing the else after return. `knownProtocol` is an `Optional<KnownProtocolKind>`, so leaving as `auto`

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  ASTContext &C = tc.Context;
+  auto parentDC = type->getDeclContext();
+  auto superType = type->getSuperclass();
+  return (bool)tc.conformsToProtocol(superType,
```
Is this bool conversion necessary?

itaiferber replied:
Shouldn't be. Removing

itaiferber replied:
I stand corrected; the return type is `Optional<swift::ProtocolConformanceRef>`, so the explicit cast is required.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+// Asserts that the Foundation module is loaded, that the given name does not
+// refer to more than one type, and that the entity with the given name refers
+// to a type instead of a different type of declaration.
+static TypeDecl* lookupFoundationTypeDecl(ASTContext &C, Identifier name) {
```
This violates Swift style. We use 'TypeDecl *'. This is a case that would be caught by clang-format.

itaiferber replied:
I was following style I'd seen earlier, but fixed throughout the file

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                     ConformanceCheckFlags::Used);
+}
+
+// Looks up a type declaration with the given name in the Foundation module.
```
If this is a doxygen comment, it should use 3 slashes ala LLVM style.

http://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments

itaiferber replied:
These weren't meant to be Doxygen comments, but no harm in documenting everything here. Added

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+// refer to more than one type, and that the entity with the given name refers
+// to a type instead of a different type of declaration.
+static TypeDecl* lookupFoundationTypeDecl(ASTContext &C, Identifier name) {
+  auto foundationModule = C.getLoadedModule(C.Id_Foundation);
```
If this is a pointer, please use 'auto *', rather than just auto.

itaiferber replied:
👍

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  auto foundationModule = C.getLoadedModule(C.Id_Foundation);
+  assert(foundationModule && "Foundation module must be loaded.");
+
+  SmallVector<ValueDecl*, 1> results;
```
'ValueDecl *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                 DeclName(name), Identifier());
+  assert(results.size() == 1 && "Ambiguous/missing type.");
+
+  auto typeDecl = dyn_cast<TypeDecl>(results[0]);
```
Another case where you should use 'auto *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+        tc.conformsToProtocol(varDecl->getType(), codableProto, context,
+                              ConformanceCheckFlags::Used)) {
+      callback(varDecl);
+    } else {
```
Please change this to a continue. In LLVM style in general, we want to ignore unnecessary else when possible.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  return allConform;
+}
+
+// Validates the given CodingKeys enum decl by ensuring its cases are a 1-to-1
```
Another doxygen-ish comment. 3 slashes? = )

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  // NOTE: If we change the behavior to ignore vars with default values, then we
+  //       can further filter out the type names to remove those which
+  //       correspond to vars with default values.
+  llvm::DenseSet<Identifier> names;
```
Do you really need to use a DenseSet here. DenseSet always allocates an internal data structure that has 64 initial elements. How big of a set do you expect to see by default? SmallDenseSet may be more appropriate.

itaiferber replied:
Wasn't aware that `DenseSet` allocated so much. I'd intended to use `SmallDenseSet` anyway but I guess I forgot to make that change. Fixed

gottesmm replied:
Yea... it surprises a lot of people. The nice thing about SmallDenseSet is that IIRC if you have a pointer, it "automagically" reduces down to a SmallPtrSet which is even more efficient.

gottesmm commented:
Now I am talking about adding this to the top of the file:

//===--- test.cpp ---------------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//


> On Mar 17, 2017, at 11:36 AM, Itai Ferber <notifications@github.com> wrote:
> 
>  commented.
> Reply to this email directly, view it o <https://github.com/apple/swift/pull/8125#discussion_r106718681>

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  llvm::DenseSet<Identifier> names;
+
+  auto storedProperties = type->getStoredProperties(/*skipInaccessible=*/true);
+  auto validVarCallback = [&names](VarDecl *varDecl) {
```
Why does this need to be out of line? Why not just make it an argument to validateVarsConformToCodable.

itaiferber replied:
The if statement becomes really unreadable if the condition has an inline lambda nested in it; it's too easy to confuse the lambda for the body of the if.

itaiferber commented:
Ah, got it — will do!

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+getOrSynthesizeCodingKeysDecl(TypeChecker &tc, Decl *parentDecl,
+                              NominalTypeDecl *type) {
+  ASTContext &C = tc.Context;
+  auto parentDC = cast<DeclContext>(parentDecl);
```
'auto *'

gottesmm replied:
As well as all of the cases in the next couple of lines that are actually pointers.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+// synthesizes a new enum based on the Codable members of the given type
+// (nullptr if unable to synthesize).
+static std::pair<NominalTypeDecl*, /* attempted to synthesize? */bool>
+getOrSynthesizeCodingKeysDecl(TypeChecker &tc, Decl *parentDecl,
```
This function is huge. It clearly has several nice subroutines in it. Can you split this up?

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  auto typeDC = cast<DeclContext>(type);
+
+  auto codingKeysDecls = type->lookupDirect(DeclName(C.Id_CodingKeys));
+  if (!codingKeysDecls.empty()) {
```
For instance, this could be refactored out into its own method.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+createKeyedContainer(ASTContext &C, DeclContext *DC, Identifier containerBase,
+                     Type keyType) {
+  // Look up Keyed*Container
+  auto keyedContainerDecl =
```
Another 'auto *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                             C.AllocateCopy(boundType));
+
+  // let container : Keyed*Container<KeyType>
+  auto containerDecl = new (C) VarDecl(/*IsStatic=*/false, /*IsLet=*/true,
```
You know what I am going to ask for here... ; ). 'auto *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  // Full bound base.container(keyedBy: CodingKeys.self) call
+  Expr* args[1] = { codingKeysMetaTypeExpr };
+  Identifier argLabels[1] = { C.Id_keyedBy };
+  return CallExpr::createImplicit(C, unboundCall, C.AllocateCopy(args),
```
Another case, where you should be using TinyPtrVector or SmallVector.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                  C.AllocateCopy(argLabels));
+}
+
+static void deriveBodyCodable_encode(AbstractFunctionDecl *encodeDecl) {
```
This is another huge function that should be split up into multiple functions. Also '_encode' needs to go. That does not match LLVM style or really anything in the Swift code base.

itaiferber replied:
This matches the style for all the other derived conformance implementations. `derive{ProtocolName}_{methodName}` for the method declaration, and `deriveBody{ProtocolName}_{methodName}` for the body synthesizer. I don't at all mind changing this to something more readable; I'd just copied the coding style from the other implementations as best as I could assuming it was the preferred way.

gottesmm replied:
If that is what is already being done in the rest of the files, you should use that. We can fix all of them together at once. It definitely is weird from an llvm style standpoint.

itaiferber replied:
Believe me, I know — I didn't enjoy propagating it. 😛  But figured that consistency was more important here than absolute correctness. I think we can do a pass on all of the derived conformances at some point to clean them all up in a way that's consistent. Some do a lot of stuff manually that was later pulled out into helper functions that were not backported, and stuff like that.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  // }
+
+  // The enclosing type decl.
+  auto typeDecl = cast<NominalTypeDecl>(encodeDecl->getDeclContext());
```
More 'auto *'. Also check out the lines below.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+
+  auto funcDC = cast<DeclContext>(encodeDecl);
+  ASTContext &C = funcDC->getASTContext();
+  auto &tc = *(TypeChecker *)C.getLazyResolver();
```
Please do not use cstyle-cases here. Please use c++ casts.

itaiferber replied:
Fixed

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+  // b) We had to derive it right here
+  assert(!derived && "Had to derive CodingKeys type too late.");
+  // c) The type is not an enum
+  auto codingKeysEnum = cast<EnumDecl>(codingKeysDecl);
```
More 'auto *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                            C.Id_KeyedEncodingContainer,
+                                            codingKeysType);
+
+  auto containerExpr = new (C) DeclRefExpr(ConcreteDeclRef(containerDecl),
```
More 'auto *'

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                           AccessSemantics::DirectToStorage);
+
+  auto enumElements = codingKeysEnum->getAllElements();
+  if (!enumElements.empty()) {
```
This can be extracted into its own function.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+    // `let container` (containerExpr) is generated above.
+
+    // encoder
+    auto encoderParam = encodeDecl->getParameterList(1)->get(0);
```
More 'auto *' everywhere in the rest of this scope.

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+
+    // Now need to generate `try container.encode(x, forKey: .x)` for all
+    // existing properties.
+    for (auto elt : enumElements) {
```
Is this a pointer? Or a reference? In these cases it is good to put in the '*' or the '&' so people down the line are able to understand the ownership semantics of the for each. We have had several cases of people assuming an auto was a & when it was really a *& resulting in miscompiles.

itaiferber replied:
Pointer; fixed

gottesmm commented on lib/Sema/DerivedConformanceCodable.cpp:
```diff
@@ -0,0 +1,950 @@
+                                                  /*Implicit=*/true);
+
+      // container.encode(self.x, forKey: CodingKeys.x)
+      Expr* args[2] = { varExpr, keyExpr };
```
Please use a SmallVector here.

gottesmm reviewed (commented):
Please fix these issues and run this through git-clang-format. Once that is done, I will look again.

gottesmm commented:
@itaiferber If you are unfamiliar with git-clang-format, it will take care of a ton of the issues I brought up, but it only does it on the git-diff of what you changed, so you don't have to worry about changing other pieces of code you didn't touch.

itaiferber commented:
@gottesmm Thanks for all of these — I will address them and re-run through `git-clang-format`.

gottesmm commented:
@itaiferber Btw, of the changes I mentioned, git-clang-format will handle the 'Type*' => 'Type *' transformations. So I would just let it handle those for you

itaiferber commented:
@gottesmm Already fixed those manually. :) Any idea why it wouldn't catch those in `DerivedCodableCodingKeys.cpp`? I have similar violations in there (`CastStmt*` → `CastStmt *`) but running the tool (`../clang/tools/clang-format/git-clang-format --binary (which clang-format) --diff -- lib/Sema/DerivedConformanceCodingKey.cpp`) reports no changes to be made. What am I missing? Adding `--style LLVM` doesn't affect it

Commit:
Integrate changes from feedback

* Add Swift headers
* Avoid else-after-return
* Convert documentation to Doxygen comments
* Follow correct pointer style
* Add missing pointer/reference annotations on auto variables
* Use SmallDenseSet instead of DenseSet
* Split up large functions (and factor out repeated code where possible)
* Avoid casting TypeChecker

itaiferber commented:
@gottesmm That last push should've address most if not all of your concerns, hopefully. Refactored a lot of code in `DerivedConformanceCodingKeys.cpp` that should shorten it up considerably. Unit tests pass as they did before, and `clang-format` has nothing to complain about but the indentation of function calls (most of which I find unreadable once it formats them)

gottesmm commented:
> On Mar 20, 2017, at 2:42 PM, Itai Ferber <notifications@github.com> wrote:
> 
> @gottesmm <https://github.com/gottesmm> That last push should've address most if not all of your concerns, hopefully. Refactored a lot of code in DerivedConformanceCodingKeys.cpp that should shorten it up considerably. Unit tests pass as they did before, and clang-format has nothing to complain about but the indentation of function calls (most of which I find unreadable once it formats them)
> 
> 

I will look at this later today, I gotta finish debugging something first. Even so, I noticed while looking at this real quick that the change in lib/AST/Decl.h I don't think has been addressed.

Michael
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift/pull/8125#issuecomment-287907432>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAee37jSzFQ-GHuhZRDrzpWvir9cnX3Iks5rnvLcgaJpZM4Memyu>.
>

Commit:
Don't decode a let var with a default value

Commit:
Split derived conformance for Codable into Encodable and Decodable

Commit:
Make CodingKey.stringValue non-optional

Commit:
Fix stringValue nil return for empty CodingKey enums

Commit:
Update CodingKey unit tests with non-optional stringValue

CodaFi commented on lib/AST/Decl.cpp:
```diff
@@ -2092,9 +2092,42 @@ bool NominalTypeDecl::derivesProtocolConformance(ProtocolDecl *protocol) const {
+      }
+
+      // Empty enums are allowed to conform as well.
+      return enumDecl->getAllElements().empty() ||
```
Just a heads up: I'm changing `hasOnlyCasesWithoutAssociatedValues` so you no longer need the first part of this clause.

itaiferber replied:
Thanks for the heads up on this. (Sorry for the delay, got back from vacation yesterday.) Is this change in, or not yet? Any good way for me to track the change so I can update accordingly?

Commit:
Sink core protocols and types into the Swift stdlib

All referenced known types are now part of the Swift stdlib, rather than Foundation

itaiferber commented:
BTW, I'm currently working on shuffling PRs (this and #8124) so that all the Swift stdlib/derived conformance stuff will be part of one PR and can land together, and the Foundation bits can land separately. **This should NOT be merged.**

itaiferber commented:
The work put up here has been subsumed into #9004.
